Bistatic delay

hfattahi opened the pull request:
This PR adds a method to the BurstSLC class to compute bistatic delay as described by [Gisinger et al 2021](https://ieeexplore.ieee.org/abstract/document/9115211). I have tried to put documentation in the code for a caveat about unclear initial bulk bistatic delay correction applied by ESA. At the moment I don't have a clear evidence if ESA has used the middle of the busrt of interest to compute their approximate bulk bistatic delay or the middle of the subswath 2. The latter has been mentioned by Gisinegr et al , but was not found in any ESA document so far.

LiangJYu commented on src/s1reader/s1_burst_slc.py:
```diff
@@ -423,6 +423,55 @@ def as_dict(self):
+        mid_range = self.starting_range + 0.5 * self.width * self.range_pixel_spacing
+        tau_mid = mid_range * 2.0 / isce3.core.speed_of_light
+
+        nx = int(self.width / xstep)
+        ny = int(self.length / ystep)
+
+        x = np.arange(0, (nx+1)*xstep, xstep, dtype = int)
+        y = np.arange(0, (ny+1)*ystep, ystep, dtype = int)
+        x_mesh = np.meshgrid(x, y)[0]
+        slant_range = self.starting_range + x_mesh * self.range_pixel_spacing
+        tau = slant_range * 2.0 / isce3.core.speed_of_light
+
+        # the first term (tau_mid/2) is the bulk bistatic delay which was
+        # removed from the orginial azimuth time by the ESA IPF. Based on
+        # Gisinger et al, 2021, ESA IPF has used the mid of the second subswath
+        # to compute the bulk bistatic delay. However currently we have not
+        # been able to verify this from ESA documents. In this implementation
+        # we have used the range to the middle of the burst of interest to
+        # compute the bulk bistatic delay. The correction can be potentially
+        # biased by the amount of discrepancy between the ESA IPF bulk bistatic
+        # correction and our assumed middle of the burst of interest.
+        bistatic_correction = tau_mid / 2 + tau / 2 - tau0
```
```suggestion
        nx = np.ceil(self.width / xstep)
        ny = np.ceil(self.length / ystep)

        x = np.arange(0, nx*xstep, xstep, dtype=int)
        y = np.arange(0, ny*ystep, ystep, dtype=int)

        slant_range = self.starting_range + x * self.range_pixel_spacing
        tau = slant_range * 2.0 / isce3.core.speed_of_light

        # the first term (tau_mid/2) is the bulk bistatic delay which was
        # removed from the orginial azimuth time by the ESA IPF. Based on
        # Gisinger et al, 2021, ESA IPF has used the mid of the second subswath
        # to compute the bulk bistatic delay. However currently we have not
        # been able to verify this from ESA documents. In this implementation
        # we have used the range to the middle of the burst of interest to
        # compute the bulk bistatic delay. The correction can be potentially
        # biased by the amount of discrepancy between the ESA IPF bulk bistatic
        # correction and our assumed middle of the burst of interest.
        bistatic_correction_vec = tau_mid / 2 + tau / 2 - tau0
        bistatic_correction = np.stack([bistatic_correction_vec for _ in range(ny)],
                                       axis=0)
```
Since all the lines are the same, this reduces amount of computation and memory usage. Also nits for no spaces in named parameter assignment in `np.arange`.

LiangJYu reviewed (approved):
LGTM

vbrancat reviewed (commented):
Thanks a lot for this PR. I am not sure if the document I posted it might be of help.

hfattahi commented on src/s1reader/s1_burst_slc.py:
```diff
@@ -423,6 +423,55 @@ def as_dict(self):
+        # we have used the range to the middle of the burst of interest to
+        # compute the bulk bistatic delay. The correction can be potentially
+        # biased by the amount of discrepancy between the ESA IPF bulk bistatic
+        # correction and our assumed middle of the burst of interest.
```
Thanks @vbrancat . Yes that is very helpful. Actually that's exactly what I needed. 

Now we need to think about an implementation strategy on how to keep track of the near range to the mid subswath for each burst (not a very good choice they made really!).  For any burst in sub-swath 1 and 2 we have that "near range of mid subswath" as the "far range" and "near range" of the burst respectively. But for a burst in subswath 3 we are not keep tracking of the near range of subswath 2 and that may need some reorganizing of the reader. Alternatively we could assume for a burst in subswath 3, the near range to subswath 2 is : 

near_range_to_subswath_2 = near_range_burst_in_swath_3 - width*rangespacing

width being the width of the busrt in subswath 3 because we don't have a burst in subswath 2. If width is off even by 500 pixels, it would only lead to ~6 cm geolocation error in azimuth which is negligible in my opinion.

hfattahi replied:
@yunjunz  pointed out [ETAD ATBD](https://sentinels.copernicus.eu/documents/247904/4629150/Sentinel-1-ETAD-Algorithm-Technical-Baseline-Document.pdf) which equation 56 indicated middle of IW2. This is inconsistent with the Guide to S1 geocoding document. Still waiting on a confirmation from folks from ESA & DLR to see which one really the ESA IPF has used. 
Probably in the meantime it is useful to be prepared and keep track of iw1, iw2, iw3 near,mid,far ranges in the reader.

vbrancat replied:
The ETAD ATBD might probably be the latest version of what they are doing. However, in my mind, ETAD and IPF processor are two different things. I could not find any latest mention to bistatic delay correction in the latest Sentinel 1 product specification document

yunjunz commented:
@hfattahi could you please do a `git rebase` for this PR so that I could also use the latest `main` branch for testing?

hfattahi commented:
> @hfattahi could you please do a `git rebase` for this PR so that I could also use the latest `main` branch for testing?
> 
> Update: I could git rebase locally myself, so it's fine to not rebase here.

I did rebase if that helps.

yunjunz commented:
Here is a comparison of one burst in **IW2**, between S1-ETAD products from ESA and this PR: they are identical.

![Screen Shot 2022-05-04 at 3 16 40 PM](https://user-images.githubusercontent.com/13143710/166834670-d7c676f5-1fb8-4a05-a875-cf3583aae5f7.png)

Here is a comparison of one burst in **IW1**, between S1-ETAD products from ESA and this PR: there is an offset of 0.00017.

![Screen Shot 2022-05-04 at 3 17 48 PM](https://user-images.githubusercontent.com/13143710/166834901-71fff0e6-5689-44bf-b95a-0b6980092006.png)

hfattahi commented:
> Here is a comparison of one burst in **IW2**, between S1-ETAD products from ESA and this PR: they are identical.
> 
> ![Screen Shot 2022-05-04 at 3 16 40 PM](https://user-images.githubusercontent.com/13143710/166834670-d7c676f5-1fb8-4a05-a875-cf3583aae5f7.png)
> 
> Here is a comparison of one burst in **IW1**, between S1-ETAD products from ESA and this PR: there is an offset of 0.00017.
> 
> ![Screen Shot 2022-05-04 at 3 17 48 PM](https://user-images.githubusercontent.com/13143710/166834901-71fff0e6-5689-44bf-b95a-0b6980092006.png)
> 
> Update: after using the IW2 `mid_range` instead of the IW1 `mid_range` for the example above, the two results match, as below:
> 
> ![Screen Shot 2022-05-04 at 5 21 02 PM](https://user-images.githubusercontent.com/13143710/166847441-3666bb54-c00a-46a0-8d2a-6c3b00a332e5.png)
> 
> So, the solution could be calculating the `iw2_mid_range` in `burst_from_xml()` and save as metadata in the burst class. To implement this, we could:
> 
> 1. guess the `iw2_annotation_path` from the input `annotation_path` inside `burst_from_xml()`, or
> 2. feed the `iw2_annotation_path` to `burst_from_xml()` as a new arguments.
> 
> Option 2) is simpler to implement and cleaner from the developer's view, while option 1) might look cleaner from the user's view. What do you think @LiangJYu @vbrancat @hfattahi ?
> 
> Note: I multiplied the s1-reader output by `-1`.

Thanks @yunjunz for doing this interesting test. At least now we are consistent with ETAD ATBD. However, we are still not sure if ETAD ATBD is correct or the ESA IPF document on using the middle of IW2 or near range of IW2. We can move forward with ETAD ATBD for now. We can revise when we hear back from ESA folks. 

I think option-2 seems cleaner to me.

yunjunz reviewed (approved):
Looks all good to me. Thank you @hfattahi for the implementation!

vbrancat reviewed (approved):
The implementation looks good to me. However, to be deadly sure we should clarify whether there is any discrepancy between the IPF implementation and what is the described in the ETAD ATBD. However, this clarification is beyond the scope of this PR.

hfattahi commented:
Thank you @yunjunz for addressing the comments from Liang and adding the mid range to IW2. Looks great.

Thank you @vbrancat. I agree. I opened this issue https://github.com/opera-adt/s1-reader/issues/35 to make sure we will get back to it when we hear from ESA on the inconsistency. 

I was about to merge. Looks like someone pushed the button already. hahaha Thank you all.
